components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -765,6 +773,169 @@ IJsonSerializer jsonSerializer
+            List<PublishedNodesEntryModel> request,
+            CancellationToken ct = default
+        ) {
+            var methodName = nameof(AddOrUpdateEndpointsAsync);
+            _logger.Information("{methodName} method triggered ... ", methodName);
+            var sw = Stopwatch.StartNew();
+
+            if (request is null) {
+                _logger.Information("{methodName} method finished in {elapsed}", methodName, sw.Elapsed);
+                sw.Stop();
+
+                throw new MethodCallStatusException((int)HttpStatusCode.BadRequest, "null request provided");
+            }
+
+            var response = new List<string>();
+
+            await _lockConfig.WaitAsync(ct).ConfigureAwait(false);
+            try {
+                // ToDo: Uncomment ValidateRequest() once our test requests pass validation.
+                // This will throw SerializerException if values of request fields are not conformant.
+                //ValidateRequest(request);
+
+                // First let's check that endpoints that we are asked to remove exist.
+                var dataSetsToRemove = request.Where(e => e.OpcNodes is null || e.OpcNodes.Count == 0).ToList();
+                var dataSetsToUpdate = request.Where(e => !(e.OpcNodes is null) && e.OpcNodes.Count > 0).ToList();
+
+                foreach (var dataSetToRemove in dataSetsToRemove) {
+                    var foundDataSet = false;
+                    foreach (var entry in _publishedNodesEntries) {
+                        foundDataSet = foundDataSet || IsSameDataSet(entry, dataSetToRemove);
+                    }
+                    if (!foundDataSet) {
+                        throw new MethodCallStatusException((int)HttpStatusCode.NotFound,
+                            $"Endpoint not found: {dataSetToRemove.EndpointUrl}");
+                    }
+                }
+
+                var existingGroups = new List<PublishedNodesEntryModel>();
+                var requestDataSetsFound = Enumerable.Repeat(false, request.Count).ToList(); ;
+
+                foreach (var entry in _publishedNodesEntries) {
+                    var groupFound = false;
+
+                    for (var k = 0; k < request.Count; ++k) {
+                        var dataSetToUpdate = request[k];
+                        if (entry.HasSameGroup(dataSetToUpdate)) {
+                            groupFound = true;
+
+                            // We may have several entries with the same DataSetGroup definition,
+                            // so we will update nodes only if the whole DataSet definition matches.
+                            if (IsSameDataSet(entry, dataSetToUpdate)) {
+                                if (dataSetToUpdate.OpcNodes is null || dataSetToUpdate.OpcNodes.Count == 0 || requestDataSetsFound[k]) {
+                                    // In this case existing OpcNodes entries should be cleaned up.
+                                    entry.OpcNodes.Clear();
```

Comment:
Let me see what will happen if we have an entry with null `OpcNodes` in JSON file.

Reply:
I've added tests to validate:
* that empty or null `OpcNodes` from published nodes JSON file do not result in entries in `_publishedNodesEntries`.
* that `null` requests or requests with `null` or empty `OpcNodes` are reported as errors.

Reply:
Also changed the call to be ` entry.OpcNodes?.Clear();` so that it does not throw exceptions in those cases. Entries with `null` `OpcNode` will be cleaned up in the next step.